src/main/java/org/kiwiproject/dropwizard/error/dao/jdk/JdbcApplicationErrorDao.java

```diff
@@ -0,0 +1,322 @@
+        try (var conn = connection();
+             var ps = conn.prepareStatement("select count(id) from application_errors where resolved = false and updated_at >= ?")) {
+
+            ps.setTimestamp(1, timestampFromZonedDateTime(since));
+
+            try (var rs = ps.executeQuery()) {
+                nextOrThrow(rs);
+                return rs.getLong(1);
+            }
+        } catch (SQLException e) {
+            throw new ApplicationErrorJdbcException(e);
+        }
+    }
+
+    @Override
+    public long countUnresolvedErrorsOnHostSince(ZonedDateTime since, String hostName, String ipAddress) {
+        try (var conn = connection();
+             var ps = conn.prepareStatement("select count(id) from application_errors" +
+                     " where resolved = false and updated_at >= ? and host_name = ? and ip_address = ?")) {
+
+            ps.setTimestamp(1, timestampFromZonedDateTime(since));
+            ps.setString(2, hostName);
+            ps.setString(3, ipAddress);
+
+            try (var rs = ps.executeQuery()) {
+                nextOrThrow(rs);
+                return rs.getLong(1);
+            }
+        } catch (SQLException e) {
+            throw new ApplicationErrorJdbcException(e);
+        }
+    }
+
+    @Override
+    public List<ApplicationError> getAllErrors(int pageNumber, int pageSize) {
+        int offset = checkPagingArgumentsAndCalculateZeroBasedOffset(pageNumber, pageSize);
+        var sql = "select * from application_errors order by updated_at desc" +
+                paginationClause(pageSize, offset);
+
+        try (var conn = connection(); var stmt = conn.createStatement(); var rs = stmt.executeQuery(sql)) {
+            return collectErrors(rs, pageSize);
+        } catch (SQLException e) {
+            throw new ApplicationErrorJdbcException(e);
+        }
+    }
+
+    @Override
+    public List<ApplicationError> getErrors(ApplicationErrorStatus status, int pageNumber, int pageSize) {
+        checkNotNull(status, "status cannot be null");
+
+        return switch (status) {
+            case ALL -> getAllErrors(pageNumber, pageSize);
+            case RESOLVED -> getErrors(true, pageNumber, pageSize);
+            case UNRESOLVED -> getErrors(false, pageNumber, pageSize);
+        };
+    }
+
+    private List<ApplicationError> getErrors(boolean resolved, int pageNumber, int pageSize) {
+        int offset = checkPagingArgumentsAndCalculateZeroBasedOffset(pageNumber, pageSize);
+        var sql = "select * from application_errors where resolved = ? order by updated_at desc"
+                + paginationClause(pageSize, offset);
+
+        try (var conn = connection(); var ps = conn.prepareStatement(sql)) {
+            ps.setBoolean(1, resolved);
+            return collectErrors(ps, pageSize);
+        } catch (SQLException e) {
+            throw new ApplicationErrorJdbcException(e);
+        }
+    }
+
+    private static String paginationClause(int pageSize, int offset) {
```

Comment:
Should this use the native SQL `OFFSET n ROWS FETCH NEXT p ONLY`syntax?

We'd also want to change the JDBI implementation too...

Reply:
Shifting this to a discussion #319